### PR TITLE
SE Development Models

### PR DESCRIPTION
**Pull request by Vaibhavpratapsingh22**
Chapter -2 created .

### CODE REVIEW

File: 4-software-engineering/2-development-models.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+Like humans when take birth they pass through certain stages like adolescence to mature than to old age . Similarly from the time of inception to the time of implementation & maintainance the software undergoes several phases . 
+
+The Systems Development Life Cycle (SDLC) gives structure to the challenges of transitioning from the beginning to the end of your project without forgetting a step.
+
+A number of different SDLC methodologies are used today to guide professionals through their project-based work.
+
+## Phases in Software development models
+
+![phases](https://ncube-digest.com/wp-content/uploads/2019/08/SDLC1.jpg)
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:11]: Image is not loading on my system

File: 4-software-engineering/2-development-models.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+* In the deployment phase, the application is made available to users. Many companies prefer to automate the deployment phase. This can be as simple as a payment portal and download link on the company website. It could also be downloading an application on a smartphone.
+
+* Deployment can also be complex. Upgrading a company-wide database to a newly-developed application is one example. Because there are several other systems used by the database, integrating the upgrade can take more time and effort.
+
+### 7. Operations and Maintenance
+* At this point, the development cycle is almost finished. The application is done and being used in the field. The Operation and Maintenance phase is still important, though. In this phase, users discover bugs that weren’t found during testing. These errors need to be resolved, which can spawn new development cycles.
+
+## Waterfall Model
+
+The waterfall model was introduced by Winston Royce in 1970 . In waterfall model all development activities are divided into various phases . The fases are organized in linear order.
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:63]:
```suggestion
The waterfall model was introduced by Winston Royce in 1970. In the waterfall model, all development activities are divided into various phases. The phases are organized in linear order. 
```

File: 4-software-engineering/2-development-models.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+The waterfall model was introduced by Winston Royce in 1970 . In waterfall model all development activities are divided into various phases . The fases are organized in linear order. 
+Therefore each phase begins only when the previous phase has finished . Each phase has well-defined starting and ending points . The output of one phase are used as the input of the second phase that's why it s also called **linear sequential model** .
+
+Various phases of waterfall model are performed one after the other and fall like water falling from the sky that's why it is known as **Waterfall Model** .
+
+A waterfall model has following phases as shown :
+
+![waterfall](https://www.scnsoft.com/blog-pictures/custom-software-development/1-waterfall-.png)
+
+### Applictions of Waterfall Model
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:72]: You have explained the phases of waterfall model under the label Applications. 
Elements explanation is important - keep that here but Applications is - where is waterfall model used.

File: 4-software-engineering/2-development-models.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+3. **Design**: This stage largely covers technical design requirements, such as programming language, data layers, services, etc. A design specification will typically be created that outlines how exactly the business logic covered in analysis will be technically implemented.
+4. **Implementation**: The actual source code is finally written in this fourth stage, implementing all models, business logic, and service integrations that were specified in the prior stages.
+5. **Testing**: During this stage, QA, beta testers, and all other testers systematically discover and report issues within the application that need to be resolved. It is not uncommon for this phase to cause a “necessary repeat” of the previous coding phase, in order for revealed bugs to be properly squashed.
+6. **Maintainance** : Once the system passes all the tests , it is deliveredd to the customer . Any modification made to the system after initial delivery are usually attributed to this phase . 
+This phase involves creating computer - compatible files , training the staff , installing hardware and telecommunication network before the system is up and running . 
+
+### Advantages
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:82]: Maybe something more relatable to the software industry. These ones are pretty straight-forward. 
for eg - It is simple one - how and why and why is it an advantage?

File: 4-software-engineering/2-development-models.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+1. It is a linear model
+2. It is simple one
+3. Easy to explain to the user
+4. Stages and activities are well - defined
+5. Widely used
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:87]: Its not widely used anymore. It was popular till late 90's and then Agile methods kicked-in

File: 4-software-engineering/2-development-models.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+### Disadvantages
+1. It is difficult to define all requirements at the beginning of a project .
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:90]:
```suggestion
1. It is difficult to define all requirements at the beginning of a project.
```

File: 4-software-engineering/2-development-models.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+2. Cycle time too long .
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:91]: None of these are sentences, therefore all full-stops can be ommitted

File: 4-software-engineering/2-development-models.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+3. Disallows changes .
+4. We cannot go in backward direction while SDLC performs .
+5. If there is any mistake or any error in any phase then we cannot make good software .
+
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:95]: Summary is not here again. If i read the document - it suddenly ends.

File: 4-software-engineering/README.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+In this chapter we will look at the basics of software engineering like :
+
+1. Definition of Software
+   * Program versus Software
+   * Software Crisis
+2. Software Characteristics
+3. Definition of Software Engineering
+4. Elements of Software Engineering
+
```
Comment (udeysingh95) [inline 4-software-engineering/README.md:17]: There is no chapter 2 discussion here.

File: 4-software-engineering/2-development-models.md (hunk 1 above)
Comment (Vaibhavpratapsingh22) [reply 4-software-engineering/2-development-models.md:11]: I am changing the image :)

File: 4-software-engineering/2-development-models.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+### Disadvantages
+1. It is difficult to define all requirements at the beginning of a project 
+2. Cycle time too long as we have to work on all the phases and ia change comes in a last step we have to start from the beginning 
+3. We cannot go in backward direction while SDLC performs 
+5. If there is any mistake or any error in any phase then we cannot make good software 
+
+
+## Summary
+In this chapter we learned about  Software Development Process Models and there different phases through which we build a software . After this we have a detailed explanation of waterfall model with the help of diagram . Then we have waterfall model application , advantages , disadvantages .
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:104]:
```suggestion
In this chapter, we learned about Software Development Process Models and different phases through which we build software. After this, we had a detailed explanation of the waterfall model with the help of a diagram . Finally, we closed with waterfall model applications, advantages, and disadvantages.
```

File: 4-software-engineering/2-development-models.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+ 
+In the next chapter we will learn about Agile Methods .
```
Comment (udeysingh95) [inline 4-software-engineering/2-development-models.md:106]:
```suggestion
In the next chapter, we will learn about Agile Methods.
```

File: 4-software-engineering/README.md (hunk 9 above)
Comment (udeysingh95) [reply 4-software-engineering/README.md:16]: This is still missing. I have made a suggestion on another branch for how this abstract could be written. You can write something similar for chapter 2 as well in the SE README file.

**Comment by udeysingh95** [comment]
https://www.atlassian.com/git/tutorials/merging-vs-rebasing
check this out to learn more about merging, reading, force-pushing

File: 4-software-engineering/README.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+Software Engineering is the field of computer science that deals with the construction of large or complex software system. The software development of this magnitude is usually done by a group of software professionals rather than the single professional. The differnce between programming and software engineering is that programming is primiarly activity while software engineering is essentially a team activity.
+
+
+## Chapter 1 - Introduction
+
+It is about the basics of software engineering. We will start with how a program is different from software. Then we will head towards the software crisis. In the software crisis, we will learn its causes, characteristics. At last, we will learn about the elements of software engineering.
+
+## Chapter 2 -  Software Development Process Models
+
+In this chapter, we learned about Software Development Process Models and different phases through which we build software. After this, we had a detailed explanation of the waterfall model with the help of a diagram . Finally, we closed with waterfall model applications, advantages, and disadvantages.
```
Comment (udeysingh95) [inline 4-software-engineering/README.md:12]:
```suggestion
It is about Software Development Process Models and different phases through which we build software. After this, the waterfall model is explained in detail with the help of a diagram . At last, we will learn about the waterfall model applications, advantages, and disadvantages.
```

File: README.md [hunk 13]
```diff
@@ -10,97 +10,76 @@ Welcome to the GitHub , Data Structure and Algorithm course here U will learn :-
...
 
 ### Dynamic Method
+                         CONTENTS
+                         
+     1. Bellma-ford algorithm  2. Flyod-warshall algorithm
 
-1. Bellma-ford algorithm
-
-2. Flyod-warshall algorithm
+### Time Complexity Using Big-O Notation
```
Comment (udeysingh95) [inline README.md:71]:
```suggestion
### Big-O Notation
```

**Pull request merged**